core/sql/exp/ExpLOBaccess.cpp

```diff
@@ -1599,203 +1627,290 @@ Ex_Lob_Error ExLob::readDataToMem(char *memAddr,
+	if (writeOperLen <= 0)
+	  {
+	    getLobGlobalHeap()->deallocateMemory(lobData);
+	    return LOB_TARGET_FILE_WRITE_ERROR;
+	  }
+	getLobGlobalHeap()->deallocateMemory(lobData);
+      }
+    else // multiple chunks to read
+      {
+	err = openCursor(getRequest()->getHandleIn(), 
+		     getRequest()->getHandleInLen());
+	if (err != LOB_OPER_OK)
+	  return err;
+	while ( srcLen > 0)
+	  {
+	    chunkSize = MINOF(srcLen, lobMaxChunkMemSize);
+	    lobData = (char *) (getLobGlobalHeap())->allocateMemory(chunkSize);
+
+	    if (lobData == NULL) 
+	      {
+		getLobGlobalHeap()->deallocateMemory(lobData);
+		return LOB_SOURCE_DATA_ALLOC_ERROR;
+	      }
+	    //handle reading the multiple chunks like a cursor
+	    err = readCursor(lobData,chunkSize, getRequest()->getHandleIn(),
+			     getRequest()->getHandleInLen(), operLen);
+
+	    if ((err != LOB_OPER_OK) || (operLen != chunkSize))
+	      {
+		getLobGlobalHeap()->deallocateMemory(lobData);
+		return err;
+	      }
+       
+	    writeOperLen += pwrite(fdDestFile, lobData, chunkSize, tgtOffset) ; 
+	    if (writeOperLen <= 0)
+	      {
+		getLobGlobalHeap()->deallocateMemory(lobData);
+		return LOB_TARGET_FILE_WRITE_ERROR;
+	      }     
+	    getLobGlobalHeap()->deallocateMemory(lobData);
+	    srcLen -= chunkSize;
+	    tgtOffset += chunkSize;     
+	  }
+	closeCursor(getRequest()->getHandleIn(), 
+		    getRequest()->getHandleInLen());
+      }
     close(fdDestFile);
-
     return LOB_OPER_OK;
 }
-Ex_Lob_Error ExLob::readDataToHdfsFile(char *tgtFileName,  Int64 offset, Int64 &operLen)
-{ 
-    Ex_Lob_Error err;
 
-    ExLobDesc desc;
-    err = getDesc(desc);
-    if (err != LOB_OPER_OK)
-      return err;
 
-    if (desc.getSize() <=0 )
-      {
-	return LOB_SOURCE_FILE_READ_ERROR;
-      }
-    char *lobData = (char *) (getLobGlobalHeap())->allocateMemory(desc.getSize());
-    if (lobData == NULL) {
-      return LOB_SOURCE_DATA_ALLOC_ERROR;
-    }
-    err = readToMem(lobData, desc.getSize(),operLen);
-    if (err != LOB_OPER_OK)
-      return err;
-    // open and write to the target file
-    int openFlags = O_WRONLY|O_CREAT;
-    hdfsFile fdTgtFile = hdfsOpenFile(fs_,tgtFileName, openFlags, 0,0,0);
-    if (fdTgtFile == NULL) {
-       return LOB_TARGET_FILE_OPEN_ERROR;
-     }
-    operLen = hdfsWrite(fs_,fdTgtFile,lobData, desc.getSize());
-    if (operLen <= 0)
-      return LOB_TARGET_FILE_WRITE_ERROR;
-	
-    if (hdfsFlush(fs_, fdTgtFile)) 
-      return LOB_DATA_FLUSH_ERROR;
-	  
-    hdfsCloseFile(fs_, fdTgtFile);
-    
-    getLobGlobalHeap()->deallocateMemory(lobData);
-    fdTgtFile = NULL;
-    /*
-#ifdef SQ_USE_HDFS
-    // tbd
-#else
-    if (pread(fdData_, data, desc.getSize(), desc.getOffset()) == -1) {
-      free(data);
-      return LOB_DATA_READ_ERROR;
+Ex_Lob_Error ExLob::readDataToHdfsFile(char *tgtFileName,  Int64 offset, Int64 size, Int64 &writeOperLen, Int64 lobMaxChunkMemLen, Int32 fileflags)
```

Comment:
I have piggybacked/overloaded an Int32 param in the ExLobsOper method which is used widely for many lob operations. FileFlags is relevant only for the read interface (for extract).  So left it as an Int32.